File: C++/ds/Queue.cpp
#include <iostream>
using namespace std;

class Queue{          //Declaring class Queue using Array
private:
    int size;
    int front;
    int rear;
    int *Q;
public:
Queue()       //Initializing all class memebers in default constructor
{
    front=-1;
    rear=-1;
    size=10;
    Q=new int[size];
}
Queue(int size)   //Initializing Array size which is given as input from Main 
{
    front=-1;
    rear=-1;
    this->size=size;
    Q=new int[this->size];
}

void Enqueue(int x){
    if(rear==size-1)          //When Array is full,then Queue is full since no space to insert new element
    {
        cout <<"Queue is full" <<endl;
    }
    else
    {
        Q[rear+1] = x;
        rear++;
    }
}

int dequeue()
{
    int x=-1;
    if(front==rear)      //when no elements present in Queue, then Queue is empty
    {
        cout <<"Queue is empty" << endl;
    }
    else{
        x = Q[front+1];
        front++;
    }
return x;
}

void display()
{
    for(int i=front+1;i<=rear;i++)
    {
        cout <<Q[i] << " ";
    }
    cout << endl;
}
};


int main()
{
    int size,num;
    cout <<"Enter size of Queue: "<< endl;
    cin >>size;
    Queue q(size);                           //Here we pass the size through parameterized constructor
    cout <<"Enter the elements into Queue: "<< endl;
    for(int i=0;i<size;i++)
    {
        cin >> num;
        q.Enqueue(num);
    }
    cout <<"Displaying Elements in Queue: " << endl;  // To display elements in the Queue
    q.display();
    cout <<"To remove element from front of the Queue" <<endl;   //Queue removes elements from front in general
    int num2;
    num2=q.dequeue();
    cout <<"Removed "<< num2 << " from Queue" << endl; // Returns the removed element from front of the Queue
    return 0;
}

File: JavaScript/ds/Queue.js
class Queue {
  constructor() {
    this.items = [];
  }

  size() {
    return this.items.length;
  }

  enqueue(element) {
    this.items.push(element);
  }

  dequeue() {
    if (this.isEmpty()) throw Error("Underflow");

    return this.items.shift();
  }

  front() {
    if (this.isEmpty()) throw Error("No elements in  the queue");

    return this.items[0];
  }

  isEmpty() {
    return this.items.length === 0;
  }

  iterate(callback = null) {
    if (callback === null) {
      let str = "";
      for (let i of this.items) {
        str += i + " ";
      }
      str.trim();
      return str;
    }

    const iterated = [];
    for (let i of this.items) {
      iterated.push(callback(i));
    }

    return iterated;
  }
}

// implementation
let arr = [10, 20, 30, 40, 50];
let queue = new Queue();

console.log("Empty size:", queue.size()); // Returns 0

console.log("Is empty?", queue.isEmpty()); // Returns true

try {
  console.log(queue.dequeue()); // Will throw an error
} catch (err) {
  console.error("Error for attemtping to dequeue empty queue:", err.message);
}

// Add items in the queue
for (let i of arr) {
  queue.enqueue(i);
}

console.log("Added values:", queue.iterate()); // Displays all elements in the queue.

console.log("Front:", queue.front()); // Returns 10
console.log("Dequeue:", queue.dequeue()); // Removes 10 from the queue.
console.log("New Front:", queue.front()); // Returns 20

console.log(
  "Iterate with callback:",
  queue.iterate((element) => element / 10)
); // Maps the queue and returns an array.
